integrated_channels/blackboard/client.py

```diff
@@ -38,3 +50,97 @@ def create_course_completion(self, user_id, payload):  # pylint: disable=unused-
         """TODO"""
+
+    def _create_session(self):
+        """
+        Will only create a new session if token expiry has been reached
+        """
+        now = datetime.utcnow()
+        if self.session is None or self.expires_at is None or now >= self.expires_at:
+            # need new session if session expired, or not initialized
+            if self.session:
+                self.session.close()
+
+            # Create a new session with a valid token
+            oauth_access_token, expires_in = self._get_oauth_access_token()
+
+            session = requests.Session()
+            session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
+            session.headers['Content-Type'] = 'application/json'
+            self.session = session
+            # expiry expected after `expires_in` seconds
+            if expires_in is not None:
+                self.expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
+
+    def _get_oauth_access_token(self):
+        """Fetch access token using refresh_token workflow from Blackboard
+
+        Returns:
+            access_token (str): the OAuth access token to access the Blackboard server
+            expires_in (int): the number of seconds after which token will expire
+        Raises:
+            HTTPError: If we received a failure response code.
+            ClientError: If an unexpected response format was received that we could not parse.
+        """
+
+        if not self.enterprise_configuration.refresh_token:
+            raise ClientError(
+                "Failed to generate oauth access token: Refresh token required.",
+                HTTPStatus.INTERNAL_SERVER_ERROR
+            )
+
+        if (not self.enterprise_configuration.blackboard_base_url
+                or not self.config.oauth_token_auth_path):
+            raise ClientError(
+                "Failed to generate oauth access token: oauth path missing from configuration.",
+                HTTPStatus.INTERNAL_SERVER_ERROR
+            )
+        auth_token_url = urljoin(
+            self.enterprise_configuration.blackboard_base_url,
+            self.config.oauth_token_auth_path,
+        )
+
+        auth_token_params = {
+            'grant_type': 'refresh_token',
+            'refresh_token': self.enterprise_configuration.refresh_token,
+        }
+
+        auth_response = requests.post(
+            auth_token_url,
+            auth_token_params,
+            headers={
+                'Authorization': self._create_auth_header(),
+                'Content-Type': 'application/x-www-form-urlencoded'
+            }
+        )
+        if auth_response.status_code >= 400:
+            raise ClientError(auth_response.text, auth_response.status_code)
+        try:
+            data = auth_response.json()
+            # do not forget to save the new refresh token otherwise subsequent requests will fail
+            self.enterprise_configuration.refresh_token = data["refresh_token"]
```

Comment:
In order to make requests you need a token, in order to get a token, you need the previous token, soooo what happens on the very first request? Given `refresh_token` is a read only field how do we get the initial token we need to get more tokens?

Reply:
yes indeed, I think similar to canvas, we expect that there is a token the very first time we use this client, populated by following the manual workflow of using redirect_uri. Once that is in place, each request gets back a new refresh_token which I then store in the db for use next time, and so forth. but yeah first time won't work unless we have authorized and gotten the first refresh_token stored.